### PR TITLE
File Upload from Server to Agent

### PR DESCRIPTION
**Pull request by twigatech**
#1 Adding the upload portion of this issue.

Added agent command to upload a file from the server to the agent. User can use agent upload <agent_id> <local_file> <target_file> to queue  the upload job.

There are some bug fixes in this that could/should be another pull request. Main one that I did was getting rid of strings.Split and using strings.Fields instead. strings.Fields allows for multiple whitespace characters between parts of a command.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -15,6 +15,8 @@ F=README.MD LICENSE data/README.MD data/agents/README.MD data/db/ data/log/READM
 F2=LICENSE
 W=Windows-x64
 L=Linux-x64
+A=Linux-arm
```
Comment (Ne0nd0g) [inline Makefile]: Nice! Thanks for adding these.

File: cmd/merlinagent/main.go [hunk 2]
```diff
@@ -35,8 +37,8 @@ import (
 	"github.com/satori/go.uuid"
 	"golang.org/x/net/http2"
 
-	"github.com/ne0nd0g/merlin/pkg/agent"
```
Comment (Ne0nd0g) [inline cmd/merlinagent/main.go]: I believe best practice is to use the full repo URL. I've updated this upstream to fix the case sensitive error where `ne0nd0g` should have been `Ne0nd0g`. Please put back in the full URL path.

File: cmd/merlinagent/main.go [hunk 3]
```diff
@@ -45,7 +47,7 @@ var debug = false
 var verbose = false
 var mRun = true
 var hostUUID = uuid.NewV4()
-var url = "https://127.0.0.1:443/"
+var url = "https://127.0.0.1:4433/"
```
Comment (Ne0nd0g) [inline cmd/merlinagent/main.go]: Default port should stay at `443`.

File: cmd/merlinserver/main.go [hunk 4]
```diff
@@ -41,9 +42,9 @@ import (
 	"github.com/satori/go.uuid"
 
 	// Merlin
-	"github.com/ne0nd0g/merlin/pkg"
```
Comment (Ne0nd0g) [inline cmd/merlinserver/main.go]: I believe best practice is to use the full repo URL. I've updated this upstream to fix the case sensitive error where ne0nd0g should have been Ne0nd0g. Please put back in the full URL path.

File: cmd/merlinserver/main.go [hunk 5]
```diff
@@ -88,7 +89,7 @@ func main() {
 
 	flag.BoolVar(&verbose, "v", false, "Enable verbose output")
 	flag.BoolVar(&debug, "debug", false, "Enable debug output")
-	port := flag.Int("p", 443, "Merlin Server Port")
+	port := flag.Int("p", 4433, "Merlin Server Port")
```
Comment (Ne0nd0g) [inline cmd/merlinserver/main.go]: Default port should stay at 443.

File: cmd/merlinagent/main.go (hunk 2 above)
Comment (Ne0nd0g) [reply cmd/merlinagent/main.go]: After thinking about it more, I see the problem with having the full repo. Kind of a chicken and egg problem. Let me research what the best practice is for this situation.

Comment (Ne0nd0g) [reply cmd/merlinagent/main.go]: It looks like it is best practice to use the full repo. You would want to `go get github.com/Ne0nd0g/merlin`, rename orgin to upstream, and add your fork as orgin.
https://blog.sgmansfield.com/2016/06/working-with-forks-in-go/
http://blog.campoy.cat/2014/03/github-and-go-forking-pull-requests-and.html

Comment (twigatech) [reply cmd/merlinagent/main.go]: I think we found the same links around the same time! Yeah, you are 100% correct in the way you do it. This is my first go fork and pull request.

File: cmd/merlinagent/main.go (hunk 3 above)
Comment (twigatech) [reply cmd/merlinagent/main.go]: Changing it back, I like to avoid having to run as root.

**Comment by twigatech** [comment]
Sorry for the garbage fire of a pull request. I knew when I deleted the vendor file it would be a bad pull request. Should be good now and thanks for the putting this on the internet and dealing with the pull request.

File: cmd/merlinagent/main.go (hunk 3 above)
Comment (Ne0nd0g) [reply cmd/merlinagent/main.go]: If it helps, you can specify the URL when running it from the command line with `merlinServer.exe -url https://127.0.0.1:4433/`

**Comment by Ne0nd0g** [comment]
@twigatech Thanks for working on the requested changes. I'll follow through on this PR in the next couple of days. I really appreciate the work you've done here.

**Comment by Ne0nd0g** [comment]
I would like to message back to the server if the file upload was a success or a failure. We need to have the agent send a message back to the server if it the action was successful or failed. What are your thoughts on adding a `type` filed to the `UploadFile` message for something like "initial" and "result"? If we go with this, we will need to set the `FileBlob` to `omitempty` so that way we can use the same message type to signal the result without having to send the file contents back. I'm open to other options too.

We need to log the initial upload command and the result.

**Comment by Ne0nd0g** [comment]
I'm thinking out loud here, but maybe we could use the `cmdResult` message type and just provide it will a stdout message if there are no errors?

**Comment by twigatech** [comment]
I am not against a specific message for results of file uploads separate from cmdResults. It is nice to take advantage of re-usable code by using cmdResult. 
I don't like having the agent send the upload file message back to the server. That would sound confusing to me. I would rather have a new message type just for the results if that was the only real option. 

I was thinking about the lack of logging in my PR. I will knock that out. I am thinking off the cuff, name, size, and sha1 of file besides the normal logging stuff of time etc. Do you have anymore ideas on what could be used in the logging?

**Comment by twigatech** [comment]
Looking at the logging again, should the uploads only exist in the agent_log.txt and not the server_log.txt?

**Pull request merged**